### PR TITLE
Don't do mempool lookups for "mempool" command without a filter

### PR DESCRIPTION
**Pull request by TheBlueMatt**

### CODE REVIEW

**Comment by pstratem** [comment]
utACK

**Comment by pstratem** [comment]
This performance regression was introduced in 319b11607f8592d7ef67ec82fa73545ad7430974

**Comment by petertodd** [comment]
utACK

**Comment by jgarzik** [comment]
utACK

**Comment by gmaxwell** [comment]
ACK. Good catch, it makes a huge performance difference with a large mempool. The problem still exists with a filter set, but it's a nice step forward. This change also makes the code simpler and clearer, which is always a bonus. I think we should take this back to 0.12 as well.

**Comment by dcousens** [comment]
utACK

**Comment by laanwj** [comment]
Posthumous ACK on the code change.
But this should have had a description and motivation in the commit message and PR description.

**Comment by luke-jr** [comment]
Agreed, I was confused as to its behaviour for a while. The description given sounds like "mempool" wouldn't work at all with non-filter connections.

**Comment by jgarzik** [comment]
Da.  The description confused several devs...

**Comment by gmaxwell** [comment]
Ah, I didn't see the potential for confusion it all. Though the description was complete and accurate, what "mempool lookup" means and is ambigious I could see it how you might extract that from the description without reading the code.

For short patches I try to read the code before the description, so as to not be biased by what it says it does and fail to see what it actually does. I would have expected with three acks from regulars that someone would have pointed out something like that earlier. Live and learn.

**Comment by dcousens** [comment]
Title made sense to me,  I can understand how the misconception could exist, but, upon reading the code it was clear that it was just moving the branching logic such that `mempool.lookup(...)` is only being done when necessary instead of pre-emptively.

File: src/main.cpp [hunk 1]
```diff
@@ -4994,12 +4994,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
...
-            if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
-            if ((pfrom->pfilter && pfrom->pfilter->IsRelevantAndUpdate(tx)) ||
-               (!pfrom->pfilter))
-                vInv.push_back(inv);
+            if (pfrom->pfilter) {
+                CTransaction tx;
+                bool fInMemPool = mempool.lookup(hash, tx);
+                if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
+                if (!pfrom->pfilter->IsRelevantAndUpdate(tx)) continue;
+            }
```
Comment (dcousens) [inline src/main.cpp]: There is 1 other change that wasn't described here,  which is `!IsRelevantAndUpdate` and `vInv.size() === MAX_INV_SZ` would previously cause a `"inv"` message to be pushed.
This does not occur now.

However, given the logic that `vInv` is only increased *if* this condition is true, means AFAIK this shouldn't be an issue.

**Pull request merged**